[3.13] GH-124567: Revert the Incremental GC in 3.13

Yhg1s opened the pull request:
Revert the Incremental GC in 3.13.

This reverts PRs #116206 (commit 15309329b65a285cb7b3071f0f08ac964b61411b), #117120 (e28477f214276db941e715eebc8cdfb96c1207d9), #117213 (8bef34f625e21886b1c64544c060e19ee2e229bf), #117422 (ddf814db744006e0f42328aa15ace97c9d8ad681), #118313 (2ba1aed596ff9e7c5e193cf990f1f20f08bbf116), #123268 (b1372e2f1afe6bd89ffaebe7383418acd8171408) and #123395 (aca65112fe3e0d2e65a73d98186ac0309a70b65d). Does not revert the raising of the first generation threshold from 700 to 2000, nor changes to C APIs (even private ones).



<!-- gh-issue-number: gh-124567 -->
* Issue: gh-124567
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--124770.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

Yhg1s committed:
Revert "[3.13] GH-117759: Document incremental GC (GH-123266) (#123395)"

This reverts commit aca65112fe3e0d2e65a73d98186ac0309a70b65d.

Yhg1s committed:
Revert "[3.13] GH-122298: Restore printing of GC stats (GH-123261) (#123268)"

This reverts commit b1372e2f1afe6bd89ffaebe7383418acd8171408.

Yhg1s committed:
Revert "gh-117657: TSAN fix race on `gstate->young.count` (#118313)"

This reverts commit 2ba1aed596ff9e7c5e193cf990f1f20f08bbf116.

Yhg1s committed:
Revert "Silence compiler warnings in gc.c (#117422)"

This reverts commit ddf814db744006e0f42328aa15ace97c9d8ad681.

Yhg1s committed:
Revert "GH-117108: Set the "old space bit" to "visited" for all young objects  (#117213)"

This reverts commit 8bef34f625e21886b1c64544c060e19ee2e229bf.

Yhg1s committed:
Revert "GH-117108: Change the size of the GC increment to about 1% of the total heap size. (GH-117120)"

This reverts commit e28477f214276db941e715eebc8cdfb96c1207d9.

Yhg1s committed:
Revert "GH-108362: Incremental Cycle GC (GH-116206)"

This reverts commit 15309329b65a285cb7b3071f0f08ac964b61411b.

Yhg1s committed:
Fix merge snafu in What's New.

Yhg1s committed:
Fix the free-threaded build.

Yhg1s committed:
Undo a few revers where there is no harm in the original versions.

Yhg1s committed:
Fix TSan build of freethreaded python.

AlexWaygood commented on Modules/gcmodule.c:
```diff
@@ -326,7 +331,7 @@ gc_get_objects_impl(PyObject *module, Py_ssize_t generation)
 
     PyInterpreterState *interp = _PyInterpreterState_GET();
-    return _PyGC_GetObjects(interp, (int)generation);
+    return _PyGC_GetObjects(interp, generation);
```
There's lots of compiler warnings on this line on Windows showing up on GitHub's "files changed" tab for this PR: 
![image](https://github.com/user-attachments/assets/1c9bfcb7-d283-4036-afdd-8b4144247e62)

Yhg1s committed:
Reinstate cast.

nascheme reviewed (approved):
LGTM.  I compared gc.c from before GH-108362 (d5ebf8b71fd18d7a1f2) and it looks like your reversion got the right bits.

markshannon commented on Doc/library/gc.rst:
```diff
@@ -40,18 +40,11 @@ The :mod:`gc` module provides the following functions:
+   With no arguments, run a full collection.  The optional argument *generation*
    may be an integer specifying which generation to collect (from 0 to 2).  A
-   :exc:`ValueError` is raised if the generation number is invalid. The sum of
+   :exc:`ValueError` is raised if the generation number  is invalid. The sum of
```
Extra whitespace (if it matters)

markshannon reviewed (approved):
One (unimportant) whitespace regression, otherwise this looks correct to me.

gpshead commented on Doc/whatsnew/3.13.rst:
```diff
@@ -509,19 +509,6 @@ Incremental garbage collection
 The cycle garbage collector is now incremental.
```
shouldn't this entire section from line 504 through the gh-108632 credit below be removed?

Yhg1s replied:
Done by including Adam's changes to the docs.

Yhg1s committed:
Update the ABI.

Yhg1s committed:
Incorporate AA Turner's doc update
(https://github.com/AA-Turner/cpython/commit/d7acbbd02f9e269075d98b8a30ab6c289ad91764)

Yhg1s committed:
Merge branch '3.13' into 3.13-revert-incremental-gc

blurb-it[bot] committed:
📜🤖 Added by blurb_it.
